cmd/oras/root/backup.go

```diff
@@ -0,0 +1,423 @@
+	if err != nil {
+		return 0, fmt.Errorf("failed to fetch content of tag %q, digest %q: %w", tag, root.Digest.String(), err)
+	}
+	var index ocispec.Index
+	if err = json.Unmarshal(manifestBytes, &index); err != nil {
+		return 0, fmt.Errorf("failed to unmarshal index for tag %q, digest %q: %w", tag, root.Digest.String(), err)
+	}
+	childrenReferrers, err := graph.RecursiveFindReferrers(ctx, target, index.Manifests, extCopyGraphOpts)
+	if err != nil {
+		return 0, fmt.Errorf("failed to count referrers for children manifests of tag %q, digest %q: %w", tag, root.Digest.String(), err)
+	}
+	referrerCount += len(childrenReferrers)
+	return referrerCount, nil
+}
+
+// finalizeBackupOutput finalizes the backup output by removing temporary directories and exporting to a tar archive if needed.
+func finalizeBackupOutput(dstRoot string, opts *backupOptions, logger logrus.FieldLogger, metadataHandler metadata.BackupHandler) (returnErr error) {
+	// Remove ingest dir for a cleaner output
+	ingestDir := filepath.Join(dstRoot, "ingest")
+	if err := os.RemoveAll(ingestDir); err != nil && !errors.Is(err, fs.ErrNotExist) {
+		logger.Debugf("failed to remove ingest directory: %v", err)
+	}
+	if opts.outputFormat != outputFormatTar {
+		// If output format is not a tar, we are done
+		return nil
+	}
+
+	// exporting the backup to a tar archive
+	if err := metadataHandler.OnTarExporting(opts.output); err != nil {
+		return err
+	}
+	tarFile, err := os.Create(opts.output)
+	if err != nil {
+		return fmt.Errorf("failed to create output file %s: %w", opts.output, err)
+	}
+	defer func() {
+		err := tarFile.Close()
+		if returnErr == nil {
+			returnErr = err
+		}
+	}()
+	if err := orasio.TarDirectory(tarFile, dstRoot); err != nil {
+		// remove the output file in case of error
+		if err := os.Remove(opts.output); err != nil && !errors.Is(err, fs.ErrNotExist) {
+			logger.Debugf("failed to remove output file %s: %v", opts.output, err)
+		}
+		return fmt.Errorf("failed to create tar archive at %s: %w", opts.output, err)
+	}
+	fi, err := os.Stat(opts.output)
+	if err != nil {
+		return fmt.Errorf("failed to stat output file %s: %w", opts.output, err)
+	}
+	return metadataHandler.OnTarExported(opts.output, fi.Size())
+}
+
+// resolveTags resolves tags to their descriptors.
+// It returns the resolved tags and their corresponding descriptors.
+func resolveTags(ctx context.Context, target oras.ReadOnlyTarget, specifiedTags []string) ([]string, []ocispec.Descriptor, error) {
+	var descs []ocispec.Descriptor
+	resolve := func(tags []string) error {
+		for _, tag := range tags {
+			desc, err := oras.Resolve(ctx, target, tag, oras.DefaultResolveOptions)
+			if err != nil {
+				return fmt.Errorf("failed to resolve tag %q: %w", tag, err)
+			}
+			descs = append(descs, desc)
+		}
+		return nil
+	}
+	if len(specifiedTags) > 0 {
+		// resolve the specified tags
+		descs = make([]ocispec.Descriptor, 0, len(specifiedTags))
```

Comment:
nit: do we really need this although it does no harm?

Reply:
I think it's more efficient to handle slice growth🤔